### PR TITLE
[spike] custom NTP provider

### PR DESCRIPTION
**Pull request by supermarin**
I haven't tested with our app yet, opening the PR for thoughts

### CODE REVIEW

File: Pod/Classes/LSClockState.h [hunk 1]
```diff
@@ -1,15 +1,15 @@
 #import <Foundation/Foundation.h>
 NS_ASSUME_NONNULL_BEGIN
 
-/// A straight port/copy of the `rl-cruntime-common` ClockState javascript prototype.
-@interface LSClockState : NSObject
+@protocol LSTimeProvider
+- (NSDate *)currentTime;
+- (SInt64)offsetInMicroseconds;
```
Comment (bhs) [inline Pod/Classes/LSClockState.h:6]: I had imagined that the time provider would only have a `currentTime` method... `LSTracer` would either have a user-specified time provider XOR a `nil` time provider, and the clock correction code would all be behind `if (!self.timeProvider) { }` guards.

File: Pod/Classes/LSTracer.m [hunk 2]
```diff
@@ -348,42 +348,55 @@ - (void)flush:(void (^)(NSError *_Nullable error))doneCallback {
...
+
+                    [self handleAPIResponse:data
+                     withOriginMicroseconds:originMicroseconds
+                                 completion:cleanupBlock
+                                      error:error];
+                }]
+     resume];
+}
+
+- (void)handleAPIResponse:(NSData *)data
```
Comment (bhs) [inline Pod/Classes/LSTracer.m:368]: this is basically just a refactor from before, right? (i.e., no real logic changes to the LSClockState updates)

**Review by bhs — COMMENTED** [review]
thanks! let me know what you think about my one thought/suggestion. I don't feel strongly about it.

File: Pod/Classes/LSClockState.h (hunk 1 above)
Comment (supermarin) [reply Pod/Classes/LSClockState.h:6]: @bensigelman agreed - I originally implemented it just as a block so the client libraries don't have to implement interfaces. I've used interface just for more structured refactoring and compiler help. Will revisit to see if we can simplify even further

File: Pod/Classes/LSTracer.m (hunk 2 above)
Comment (supermarin) [reply Pod/Classes/LSTracer.m:368]: right, just moved since the upper method was getting hard to follow with added branching

File: Pod/Classes/LSClockState.h (hunk 1 above)
Comment (supermarin) [reply Pod/Classes/LSClockState.h:6]: One problem is that API requires it here:
https://github.com/supermarin/lightstep-tracer-objc/blob/1772146ddbb6e990780504deeb02f0b8a2d5bb8a/Pod/Classes/LSTracer.m#L317

Is the parameter required or optional?

Comment (bhs) [reply Pod/Classes/LSClockState.h:6]: optional -- good question.
